File: backend/app/api/v1/endpoints/blogs.py
from typing import List

from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session

from app.api.deps import get_db, get_current_director
from app.models.blog import BlogPost
from app.schemas.blog import BlogPostCreate, BlogPostRead, BlogPostUpdate

router = APIRouter()


@router.get("", response_model=List[BlogPostRead])
async def get_all_blogs(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=100),
    hot: bool = None,
    db: Session = Depends(get_db),
):
    """
    Get all blog posts.
    Matches frontend: blogService.getAll()
    """
    query = db.query(BlogPost)
    if hot is not None:
        query = query.filter(BlogPost.hot == hot)
    blogs = query.order_by(BlogPost.date.desc()).offset(skip).limit(limit).all()
    return blogs


@router.get("/{blog_id}", response_model=BlogPostRead)
async def get_blog(blog_id: int, db: Session = Depends(get_db)):
    """
    Get blog post by ID.
    Matches frontend: blogService.getById(id)
    """
    blog = db.query(BlogPost).filter(BlogPost.id == blog_id).first()
    if not blog:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Blog post not found",
        )
    return blog


@router.post("", response_model=BlogPostRead, status_code=status.HTTP_201_CREATED)
async def create_blog(
    blog_in: BlogPostCreate,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_director),
):
    """
    Create a new blog post (Director only).
    Matches frontend: blogService.create(data)
    """
    blog = BlogPost(
        title=blog_in.title,
        image=blog_in.image,
        letter=blog_in.letter,
        date=blog_in.date,
        hot=blog_in.hot,
    )
    db.add(blog)
    db.commit()
    db.refresh(blog)
    return blog


@router.put("/{blog_id}", response_model=BlogPostRead)
async def update_blog(
    blog_id: int,
    blog_in: BlogPostUpdate,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_director),
):
    """
    Update a blog post (Director only).
    Matches frontend: blogService.update(id, data)
    """
    blog = db.query(BlogPost).filter(BlogPost.id == blog_id).first()
    if not blog:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Blog post not found",
        )

    update_data = blog_in.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(blog, field, value)

    db.commit()
    db.refresh(blog)
    return blog


@router.delete("/{blog_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_blog(
    blog_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_director),
):
    """
    Delete a blog post (Director only).
    Matches frontend: blogService.delete(id)
    """
    blog = db.query(BlogPost).filter(BlogPost.id == blog_id).first()
    if not blog:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Blog post not found",
        )

    db.delete(blog)
    db.commit()
    return None

File: backend/src/routes/v1/blogs.ts
import type { FastifyPluginAsync } from 'fastify';

import { DIRECTOR_PLUS_ROLES } from '../../lib/auth.js';
import { notFound } from '../../lib/errors.js';
import { prisma } from '../../lib/prisma.js';
import { serializeBlogPost } from '../../lib/serializers.js';
import { parseDateOnly } from '../../lib/time.js';

const blogsRoutes: FastifyPluginAsync = async (fastify) => {
  fastify.get('', async (request) => {
    const query = request.query as { skip?: string | number; limit?: string | number; hot?: string | boolean };
    const skip = Math.max(Number(query.skip ?? 0), 0);
    const limit = Math.min(Math.max(Number(query.limit ?? 100), 1), 100);

    const where =
      query.hot === undefined
        ? undefined
        : { hot: query.hot === true || query.hot === 'true' || query.hot === '1' };

    const blogs = await prisma.blogPost.findMany({
      where,
      skip,
      take: limit,
      orderBy: { date: 'desc' },
    });

    return blogs.map(serializeBlogPost);
  });

  fastify.get('/:blog_id', async (request) => {
    const params = request.params as { blog_id: string };
    const blogId = Number(params.blog_id);

    const blog = await prisma.blogPost.findUnique({ where: { id: blogId } });
    if (!blog) {
      notFound('Blog post not found');
    }

    return serializeBlogPost(blog);
  });

  fastify.post('', { preHandler: [fastify.requireRoles(DIRECTOR_PLUS_ROLES, 'Director access required')] }, async (request, reply) => {
    const body = request.body as {
      title: string;
      image?: string | null;
      letter: string;
      date: string;
      hot?: boolean;
    };

    const blog = await prisma.blogPost.create({
      data: {
        title: body.title,
        image: body.image ?? null,
        letter: body.letter,
        date: parseDateOnly(body.date),
        hot: body.hot ?? false,
        created_at: new Date(),
      },
    });

    return reply.status(201).send(serializeBlogPost(blog));
  });

  fastify.put(
    '/:blog_id',
    { preHandler: [fastify.requireRoles(DIRECTOR_PLUS_ROLES, 'Director access required')] },
    async (request) => {
      const params = request.params as { blog_id: string };
      const body = request.body as {
        title?: string;
        image?: string | null;
        letter?: string;
        date?: string;
        hot?: boolean;
      };

      const blogId = Number(params.blog_id);
      const existing = await prisma.blogPost.findUnique({ where: { id: blogId } });
      if (!existing) {
        notFound('Blog post not found');
      }

      const blog = await prisma.blogPost.update({
        where: { id: blogId },
        data: {
          ...(body.title !== undefined ? { title: body.title } : {}),
          ...(body.image !== undefined ? { image: body.image } : {}),
          ...(body.letter !== undefined ? { letter: body.letter } : {}),
          ...(body.date !== undefined ? { date: parseDateOnly(body.date) } : {}),
          ...(body.hot !== undefined ? { hot: body.hot } : {}),
        },
      });

      return serializeBlogPost(blog);
    }
  );

  fastify.delete(
    '/:blog_id',
    { preHandler: [fastify.requireRoles(DIRECTOR_PLUS_ROLES, 'Director access required')] },
    async (request, reply) => {
      const params = request.params as { blog_id: string };
      const blogId = Number(params.blog_id);

      const blog = await prisma.blogPost.findUnique({ where: { id: blogId } });
      if (!blog) {
        notFound('Blog post not found');
      }

      await prisma.blogPost.delete({ where: { id: blogId } });
      return reply.status(204).send();
    }
  );
};

export default blogsRoutes;
